js/src/builtin/streams/ReadableByteStreamController.cpp

```diff
@@ -761,15 +567,231 @@ enum BYOBRequestSlots {
+              cx, args, "get desiredSize"));
+  if (!unwrappedController) {
+    return false;
+  }
+
+  // 3.10.8. ReadableByteStreamControllerGetDesiredSize, steps 1-4.
+  // 3.10.8. Step 1: Let stream be controller.[[controlledReadableStream]].
+  ReadableStream* unwrappedStream = unwrappedController->stream();
+
+  // 3.10.8. Step 2: Let state be stream.[[state]].
+  // 3.10.8. Step 3: If state is "errored", return null.
+  if (unwrappedStream->errored()) {
+    args.rval().setNull();
+    return true;
+  }
+
+  // 3.10.8. Step 4: If state is "closed", return 0.
+  if (unwrappedStream->closed()) {
+    args.rval().setInt32(0);
+    return true;
+  }
+
+  // Step 2: Return ! ReadableByteStreamControllerGetDesiredSize(this).
+  args.rval().setNumber(
+      ReadableStreamControllerGetDesiredSizeUnchecked(unwrappedController));
+  return true;
+}
+
+/**
+ * Streams spec, 3.9.4.2 close()
+ */
+static bool ReadableByteStreamController_close(JSContext* cx, unsigned argc,
+                                               Value* vp) {
+  // Step 1: If ! IsReadableByteStreamController(this) is false, throw a
+  //         TypeError exception.
+  CallArgs args = CallArgsFromVp(argc, vp);
+  Rooted<ReadableByteStreamController*> unwrappedController(
+      cx,
+      UnwrapAndTypeCheckThis<ReadableByteStreamController>(cx, args, "close"));
+  if (!unwrappedController) {
+    return false;
+  }
+
+  // Step 2: If ! ReadableByteStreamControllerCanCloseOrEnqueue(this) is
+  //         false, throw a TypeError exception.
+  if (!js::CheckReadableStreamControllerCanCloseOrEnqueue(
+          cx, unwrappedController, "close")) {
+    return false;
+  }
+
+  // Step 3: Perform ! ReadableByteStreamControllerClose(this).
+  if (!ReadableByteStreamControllerClose(cx, unwrappedController)) {
+    return false;
+  }
+
+  args.rval().setUndefined();
+  return true;
+}
+
+/**
+ * Streams spec, 3.9.4.3. enqueue ( chunk )
+ */
+static bool ReadableByteStreamController_enqueue(JSContext* cx, unsigned argc,
+                                                 Value* vp) {
+  // Step 1: If ! IsReadableByteStreamController(this) is false, throw a
+  //         TypeError exception.
+  CallArgs args = CallArgsFromVp(argc, vp);
+  Rooted<ReadableByteStreamController*> unwrappedController(
+      cx, UnwrapAndTypeCheckThis<ReadableByteStreamController>(cx, args,
+                                                               "enqueue"));
+  if (!unwrappedController) {
+    return false;
+  }
+
+  // Step 2: If ! ReadableByteStreamControllerCanCloseOrEnqueue(this) is
+  //         false, throw a TypeError exception.
+  if (!js::CheckReadableStreamControllerCanCloseOrEnqueue(
+          cx, unwrappedController, "enqueue")) {
+    return false;
+  }
+
+  // Step 3: Return ! ReadableByteStreamControllerEnqueue(this, chunk).
+  if (!ReadableByteStreamControllerEnqueue(cx, unwrappedController,
+                                           args.get(0))) {
```

Comment:
Should we be checking that this argument was supplied?
Something like:
```
if (!args.requireAtLeast(cx, "ReadableStreamDefaultControllerEnqueue", 1)) {
    return false;
}
```

Reply:
Good catch, there are a number of error paths I missed here working from the older spec. Strictly speaking we should probably be updating the `DefaultStream` implementation to do similar checks as well.